File: 1061-lexicographically-smallest-equivalent-string/1061-lexicographically-smallest-equivalent-string.java
class Solution {
    int[] parent = new int[26];

    private int findUltPar(int x) {
        if(parent[x] != x) {
            parent[x] = findUltPar(parent[x]);
        }
        return parent[x];
    }

    private void union(int u, int v) {
        int ultPar_u = findUltPar(u);
        int ultPar_v = findUltPar(v);

        if(ultPar_u != ultPar_v) {
            if(ultPar_u < ultPar_v) {
                parent[ultPar_v] = ultPar_u;
            } else {
                parent[ultPar_u] = ultPar_v;
            }
        } 
    }
    public String smallestEquivalentString(String s1, String s2, String baseStr) {
        StringBuilder sb = new StringBuilder();
        for(int i=0; i<26; i++) {
            parent[i] = i;
        }

        for(int i=0; i<s1.length(); i++) {
            int u = s1.charAt(i) - 'a';
            int v = s2.charAt(i) - 'a';

            union(u, v);
        }

        for(int i=0; i<baseStr.length(); i++) {
            int mappedChar = findUltPar(baseStr.charAt(i) - 'a');
            sb.append((char)(mappedChar + 'a'));
        }

        return sb.toString();
        }
}

File: 1061-lexicographically-smallest-equivalent-string/1061-lexicographically-smallest-equivalent-string.cpp
class Solution {
public:
    class DSU{

        vector<int> parent;
        vector<int> rank;

        public:
        DSU(int n){
            parent.resize(n);
            rank.resize(n);

            for(int i = 0; i < n; i++){
                parent[i] = i;
            }
        }

        int findPar(int u){

            if(parent[u] != u){
                parent[u] = findPar(parent[u]);
            }

            return parent[u];
        }

        void unionByRank(int u , int v){

            int par_u = findPar(u);
            int par_v = findPar(v);

            if(par_u == par_v) return;

            if(rank[par_u] < rank[par_v]){
                parent[par_u] = par_v;
            }else if(rank[par_u] < rank[par_v]){
                parent[par_v] = par_u;
            }else{
                parent[par_u] = par_v;
                rank[par_u]++;
            }
        }

        void unionByLex(int u, int v){

            int par_u = findPar(u);
            int par_v = findPar(v);

            if(par_u == par_v) return;

            if(par_u > par_v){
                parent[par_u] = par_v;
            }else{
                parent[par_v] = par_u;
            }
        }
    };
    string smallestEquivalentString(string s1, string s2, string baseStr) {
        

        int n = s1.size();

        vector<pair<int, int>> vp;

        for(int i = 0; i < n; i++){

            pair<int, int> temp;
            
            int c1 = s1[i] - 'a';
            int c2 = s2[i] - 'a';

            if(c1 != c2){
                if(c1 > c2) temp = {c2, c1};
                else temp = {c1, c2};
                vp.push_back(temp);
            }
        }

        sort(vp.begin(), vp.end());

        DSU dsu = DSU(26);

        for(auto &pr : vp){
            
            int u = pr.first;
            int v = pr.second;

            if(dsu.findPar(u) != dsu.findPar(v)){
                dsu.unionByLex(u, v);
            }
        } 

        for(int i = 0; i < baseStr.size(); i++){

            int k = (baseStr[i] - 'a');

            int par = dsu.findPar(k);
            baseStr[i] = par + 'a';
        }

        return baseStr;
    }
};